TimeOfFlight/src/TOFUtils.cc

```diff
@@ -2,145 +2,363 @@
-
-    
-    return diff.cross( unitDir ).r() ;
-    
-  }
- 
-
-  CaloHitDataVec findHitsClosestToLine( const CaloHitLayerMap& layerMap ){
-    
-    CaloHitDataVec hitVec ;
-    
-    for( auto m : layerMap ){
-      
-      const CaloHitDataVec& chv = m.second ;
-      
-      CaloHitData* closestHit =
-	*std::min_element( chv.begin() , chv.end () ,
-			   [](CaloHitData* c0, CaloHitData* c1 ){ return c0->distancefromStraightline < c1->distancefromStraightline  ; }
-	  ) ; 
-
-      hitVec.push_back( closestHit ) ;
-    }
 
-    return hitVec ;
-  }
+double TOFUtils::getTofFrankAvg( std::vector<EVENT::CalorimeterHit*> selectedHits, EVENT::Track* track, double timeResolution){
+    const TrackState* tsEcal = track->getTrackState(TrackState::AtCalorimeter);
+    Vector3D trackPosAtEcal ( tsEcal->getReferencePoint() );
 
+    int nHits = selectedHits.size();
+    if (nHits == 0) return 0.;
 
-  std::pair<float,float> computeTOFEstimator( const CaloHitDataVec& chv ){
+    double tof = 0.;
+    for ( auto hit : selectedHits ){
+        Vector3D hitPos( hit->getPosition() );
+        double dToTrack = (hitPos - trackPosAtEcal).r();
+        tof += RandGauss::shoot(hit->getTime(), timeResolution) - dToTrack/CLHEP::c_light;
+    }
+    return tof/nHits;
+}
 
-    const static float c_mm_per_ns = 299.792458 ;
 
-    int   nHit = 0 ;
-    float mean = 0. ;
-    float meansq = 0. ;
+double TOFUtils::getTofFrankFit( std::vector<EVENT::CalorimeterHit*> selectedHits, EVENT::Track* track, double timeResolution){
+    const TrackState* tsEcal = track->getTrackState(TrackState::AtCalorimeter);
+    Vector3D trackPosAtEcal ( tsEcal->getReferencePoint() );
 
-    for( auto ch : chv ){
-      float t = ch->smearedTime - ch->distanceFromReferencePoint / c_mm_per_ns ; 
-      mean   += t   ;
-      ++nHit ;
+    int nHits = selectedHits.size();
+    if (nHits == 0) return 0.;
+    else if (nHits == 1){
+        //we can't fit 1 point, but lets return something reasonable
+        Vector3D hitPos( selectedHits[0]->getPosition() );
+        double dToTrack = (hitPos - trackPosAtEcal).r();
+        return RandGauss::shoot(selectedHits[0]->getTime(), timeResolution) - dToTrack/CLHEP::c_light;
     }
-    mean /= nHit ;
 
-    for( auto ch : chv ){
-      float t = ch->smearedTime - ch->distanceFromReferencePoint / c_mm_per_ns ; 
-      meansq += ( t - mean ) * ( t - mean )   ;
+    vector <double> x, y, xErr, yErr;
+    for ( auto hit : selectedHits ){
+        Vector3D hitPos( hit->getPosition() );
+        double dToTrack = (hitPos - trackPosAtEcal).r();
+        x.push_back(dToTrack);
+        double time = RandGauss::shoot(hit->getTime(), timeResolution);
+        y.push_back(time);
+        xErr.push_back(0.);
+        yErr.push_back(0.3);
+        //OPTIMIZE: setting this to 0 is not good for the fit.. So I put random 300ps...
+        //Changing this doesn't seem to affect results, although one may want to check it more carefully
     }
```

Comment:
Yes, you are right, I will change this..

I was just careful, because having error bars do actually affect `chi2` and fit parameter errors, if we plan to use them somewhere later on for the debugging or something else. But then we need to put some proper error values not `0.3 ns`...

Here just for the reference, I tried to do two identical fits of manual data with `TGraph` and `TGrapheErrors`:

![image](https://user-images.githubusercontent.com/28848011/136179895-f4f1020a-991c-429e-aa43-cd13b04dea26.png)

The outputs are:

```
****************************************
Minimizer is Linear
Chi2                      =      691.747
NDf                       =           28
p0                        =      5.27038   +/-   1.77049     
p1                        =      2.07006   +/-   0.104844    

****************************************
Minimizer is Linear
Chi2                      =      27.6699
NDf                       =           28
p0                        =      5.27038   +/-   1.78102     
p1                        =      2.07006   +/-   0.105468   
```
You can see that `Chi2` is very different and parameter errors also can differ.

But the `p0` is unaffected, which I wasn't sure in the beginning